src/GraphQL/Internal/Name.hs

```diff
@@ -17,13 +18,67 @@ module GraphQL.Internal.Name
 
 import Protolude
 
+import qualified Data.Aeson as Aeson
 import GHC.TypeLits (Symbol, KnownSymbol, symbolVal)
-import GraphQL.Internal.Syntax.AST
-  ( Name(..)
-  , NameError(..)
-  , unsafeMakeName
-  , makeName
-  )
+import Data.Char (isDigit)
+import Data.Text as T (Text)
+import qualified Data.Attoparsec.Text as A
+import Test.QuickCheck (Arbitrary(..), elements, listOf)
+import Data.String (IsString(..))
+import Data.Text as T (Text, append, empty)
+
+import GraphQL.Internal.Syntax.Tokens (tok)
+
+-- * Name
+
+-- | A name in GraphQL.
+--
+-- https://facebook.github.io/graphql/#sec-Names
+newtype Name = Name { unName :: T.Text } deriving (Eq, Ord, Show)
+
+-- | Allow Name to be parsed with `optempty`
+--
+-- Example: node = AST.Node <$> optempty nameParser
+-- I.e. If nameParser fails, the Name field of AST.Node is set
+-- mempty rather than propagating a failure. 
+instance Monoid Name where
+    mempty  = Name T.empty
+    mappend (Name a1) (Name a2) = Name (T.append a1 a2)
```

Comment:
Rather than importing `T.empty` and `T.append`, I'd suggest just using `mempty` and `mempty`, e.g.

```haskell
instance Monoid Name where
  mempty = Name mempty
  mappend (Name a1) (Name a2) = Name (mappend a1 a2)
```

Looking at that, if you enable the `GeneralizedNewtypeDeriving` (sp?) language option, you could just change `Name` to derive `Monoid` to get the same definition.

Reply:
Yes, this was right. The `GeneralizedNewtypeDeriving` with derived `Monoid` obsoleted this code.